esphome/components/pm2005/pm2005.cpp

```diff
@@ -0,0 +1,121 @@
+#include "esphome/core/log.h"
+#include "pm2005.h"
+
+namespace esphome {
+namespace pm2005 {
+
+static const char *const TAG = "pm2005";
+
+#ifdef PM2005_USE_TYPE_2005
+static const uint8_t SITUATION_VALUE_INDEX = 3;
+static const uint8_t PM_1_0_VALUE_INDEX = 4;
+static const uint8_t PM_2_5_VALUE_INDEX = 6;
+static const uint8_t PM_10_0_VALUE_INDEX = 8;
+static const uint8_t MEASURING_VALUE_INDEX = 10;
+#else
+static const uint8_t SITUATION_VALUE_INDEX = 2;
+static const uint8_t PM_1_0_VALUE_INDEX = 3;
+static const uint8_t PM_2_5_VALUE_INDEX = 5;
+static const uint8_t PM_10_0_VALUE_INDEX = 7;
+static const uint8_t MEASURING_VALUE_INDEX = 9;
+#endif
+
+// Converts a sensor situation to a human readable string
+static const LogString *pm2005_get_situation_string(int status) {
+  switch (status) {
+    case 1:
+      return LOG_STR("Close.");
+    case 2:
+      return LOG_STR("Malfunction.");
+    case 3:
+      return LOG_STR("Under detecting.");
+    case 0x80:
+      return LOG_STR("Detecting completed.");
+    default:
+      return LOG_STR("Invalid");
+  }
+}
+
+// Converts a sensor measuring mode to a human readable string
+static const LogString *pm2005_get_measuring_mode_string(int status) {
+  switch (status) {
+    case 2:
+      return LOG_STR("Single measuring mode.");
+    case 3:
+      return LOG_STR("Continuous measuring mode.");
+    case 5:
+      return LOG_STR("Dynamic measuring mode.");
+    default:
+      return LOG_STR("Unknown");
+  }
+}
+
+void PM2005Component::update() {
+  if (this->read(data_buffer_, 12) != i2c::ERROR_OK) {
+    ESP_LOGW(TAG, "Read result failed");
+    this->status_set_warning();
+    return;
+  }
+
+  if (this->sensor_situation_ != data_buffer_[SITUATION_VALUE_INDEX]) {
+    this->sensor_situation_ = data_buffer_[SITUATION_VALUE_INDEX];
+    if (this->sensor_situation_ == 1)
+      ESP_LOGD(TAG, "Sensor situation: %s.", LOG_STR_ARG(pm2005_get_situation_string(this->sensor_situation_)));
+    else if (this->sensor_situation_ == 2) {
+      ESP_LOGD(TAG, "Sensor situation: %s.", LOG_STR_ARG(pm2005_get_situation_string(this->sensor_situation_)));
+      this->status_set_warning();
+    } else if (this->sensor_situation_ == 3)
+      ESP_LOGD(TAG, "Sensor situation: %s.", LOG_STR_ARG(pm2005_get_situation_string(this->sensor_situation_)));
+    else if (this->sensor_situation_ == 0x80) {
+      ESP_LOGD(TAG, "Sensor situation: %s.", LOG_STR_ARG(pm2005_get_situation_string(this->sensor_situation_)));
```

Comment:
My idea to use LogString was to avoid multiple log messages with repeated string like `Sensor situation`.

Couldnt you just do something like this?
```suggestion
    ESP_LOGD(TAG, "Sensor situation: %s.", LOG_STR_ARG(pm2005_get_situation_string(this->sensor_situation_)));
    if (this->sensor_situation_ == 0x80) {
```

Reply:
Then there is a big chance that we will get an unknown status in case of undocumented values, and lose the warning status in case of receiving the value 2.
I am currently checking a similar version of the code, if it works I will update it.